nixos/modules/services/networking/headscale.nix

```diff
@@ -0,0 +1,486 @@
+            '';
+          };
+          httpListen = mkOption {
+            type = types.nullOr types.str;
+            default = ":http";
+            description = ''
+              When HTTP-01 challenge is chosen, letsencrypt must set up a
+              verification endpoint, and it will be listning on:
+              :http = port 80
+            '';
+          };
+        };
+
+        certFile = mkOption {
+          type = types.nullOr types.path;
+          default = "";
+          description = ''
+            Path to already created certificate.
+          '';
+        };
+        keyFile = mkOption {
+          type = types.nullOr types.path;
+          default = "";
+          description = ''
+            Path to key for already created certificate.
+          '';
+        };
+      };
+
+      aclPolicyFile = mkOption {
+        type = types.nullOr types.path;
+        default = "";
+        description = ''
+          Path to a file containg ACL policies
+        '';
+      };
+
+    };
+
+  };
+  config = mkIf cfg.enable {
+    # Setup the headscale configuration in a known path in /etc to
+    # allow both the Server and the Client use it to find the socket
+    # for communication.
+    environment.etc."headscale/config.yaml".source = configFile;
+
+    users.groups = optionalAttrs (cfg.group == "headscale") {
+      headscale = { };
+    };
+
+    users.users = optionalAttrs (cfg.user == "headscale") {
+      headscale = {
+        description = "headscale user";
+        createHome = true;
+        home = cfg.dataDir;
+        group = cfg.group;
+        useDefaultShell = true;
+        isNormalUser = true;
+      };
+    };
+
+    systemd.services.headscale = {
+      description = "headscale coordination server for Tailscale";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      script = ''
+        ${optionalString (cfg.database.passwordFile != null) ''
+        HEADSCALE_DB_PASS="$(head -n1 ${cfg.database.passwordFile})"
+        ''}
+
+        HEADSCALE_OIDC_CLIENT_SECRET="$(head -n1 ${cfg.openIdConnect.clientSecretFile})"
+        ${cfg.package}/bin/headscale serve
+      '';
+
+      environment = {
+        # Make sure the systemd configuration reloads if the
+        # configuration has changed
+        HEADSCALE_CONFIG_SHA512 = builtins.hashFile "sha512" configFile;
+      };
+
+      serviceConfig =
+        let
+          capabilityBoundingSet = [ "CAP_CHOWN" ] ++ optional (cfg.port < 1024) "CAP_NET_BIND_SERVICE";
+        in
+        {
+          Restart = "always";
+          Type = "simple";
+          User = cfg.user;
+          Group = cfg.group;
+          WorkingDirectory = cfg.dataDir;
+
+          # Hardening options
+
+          RuntimeDirectory = mkIf (cfg.runDir == "/run/headscale") "headscale";
+          # Allow headscale group access so users can be added and use the CLI.
+          RuntimeDirectoryMode = "0770";
+
+          StateDirectory = cfg.dataDir;
```

Comment:
does this work as intended? systemd expects a relative name here. (it might even be a good idea to remove the `runDir` and `dataDir` options in favor of using `"headscale"` for both unconditionally)

Reply:
A big 👍 for that. Maybe we can even run this service as `DynamicUser`? Or not 🤷‍♂️

Reply:
I have tried to read up on `DynamicUser`, and it _might_ be possible, but for the same reason as my `/etc` config linking and access to the socket, I am not sure if I can?

Reply:
Maybe I'll poke at `DynamicUser` if I have some free time.